src/aspire/reconstruction/mean.py

```diff
@@ -1,17 +1,262 @@
+                        i, min(self.src.n, i + self.batch_size), dtype=int
+                    )
+                    pts_rot = rotated_grids(
+                        self.src.L, self.src.rotations[_range, :, :]
+                    )
+                    weights = sq_filters_f[:, :, _range]
+                    weights *= self.src.amplitudes[_range] ** 2
+
+                    if self.src.L % 2 == 0:
+                        weights[0, :, :] = 0
+                        weights[:, 0, :] = 0
+
+                    weights *= (
+                        self.weights[_range, j] * self.weights[_range, k]
+                    ).reshape(1, 1, len(_range))
+
+                    pts_rot = pts_rot.reshape((3, -1))
+                    weights = np.transpose(weights, (2, 0, 1)).flatten()
+
+                    batch_kernel = (
+                        1
+                        / (2 * self.src.L**4)
+                        * anufft(weights, pts_rot[::-1], (_2L, _2L, _2L), real=True)
+                    )
+                    kernel[k, j] += batch_kernel
+
+                    # r x r symmetric
+                    # accumulate batch entries of kernel[k,j] to kernel[j,k]
+                    if j != k:
+                        kernel[j, k] += batch_kernel
+
+        kermat_f = np.zeros((self.r, self.r, _2L, _2L, _2L))
+        logger.info("Computing non-centered Fourier Transform Kernel Mat")
+        for k in range(self.r):
+            for j in range(self.r):
+                # Ensure symmetric kernel
+                kernel[k, j, 0, :, :] = 0
+                kernel[k, j, :, 0, :] = 0
+                kernel[k, j, :, :, 0] = 0
+
+                kernel[k, j] = fft.mdim_ifftshift(kernel[k, j], range(0, 3))
+                kernel_f = fft.fftn(kernel[k, j], axes=(0, 1, 2))
+
+                kernel_f = np.real(kernel_f)
+                kermat_f[k, j] = kernel_f
+
+        return FourierKernelMat(kermat_f, centered=False)
+
+    def src_backward(self):
+        """
+        Apply adjoint mapping to source
+
+        :return: The adjoint mapping applied to the images, averaged over the whole dataset and expressed
+            as coefficients of `basis`.
+        """
+
+        # src_vols_wt_backward
+        mean_b = Volume(
+            np.zeros((self.r, self.src.L, self.src.L, self.src.L), dtype=self.dtype)
+        )
+
+        for k in range(self.r):
+            for i in range(0, self.src.n, self.batch_size):
+                im = self.src.images[i : i + self.batch_size]
+
+                batch_mean_b = (
+                    self.src.im_backward(im, i, self.weights[:, k]) / self.src.n
+                )
+                mean_b[k] += batch_mean_b.astype(self.dtype)
+
+        res = np.sqrt(self.src.n) * self.basis.evaluate_t(mean_b)
+        logger.info(f"Determined weighted adjoint mappings. Shape = {res.shape}")
+
+        return res
+
+    def conj_grad(self, b_coeff, tol=1e-5, regularizer=0):
+        n = b_coeff.shape[-1]  # 0???
```

Comment:
Please elaborate?

Reply:
Bad comment oversight on my part.  When I walked through the original code, this `shape[0]` caught me offguard.  I later learned it was being squeezed somewhere else.

More specifically, the old (`develop`) code had `np.squeeze` going into `conj_grad`. For the old code, squeeze converted the `b_coef` column to a vector of length `count`, so take shape of `0` worked.  The current code is `(r , basis.count)`, so we would want `1` or `-1`.  The later happened to work with the old code as well. Made a update to the comment.

Two of the unit tests `testOptimize1` and `testOptimize2`, were still passing a 1D vector into `conj_grad`, so I changed them to also be consistent.